src/lib/async.pl

```diff
@@ -0,0 +1,366 @@
+    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+
+*/
+
+
+:- module(async, [async_event_loop/1, await/1]).
+:- use_module(library(queues)).
+:- use_module(library(cont)).
+
+
+/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+   Public API
+- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
+
+%% async_event_loop(+Goals)
+% Goals is a list of goals that may be asyncronous, meaning that that `await/1`
+% will unify within this context and fail otherwise.
+
+async_event_loop([]).
+async_event_loop([X|Xs]) :-
+        list_queue([X|Xs],Q),
+        run_task_q(Q).
+
+%% await(+Goal)
+% When invoked within an `async_event_loop/1` call as an ancestor,
+% the `Goal` will operate as a coroutine. Note that it is asyncronous
+% but NOT concurrent or parallel without invoking an external process.
+
+await(G) :-
+        shift(zurück(G)).
+
+/** Async Programming for Scryer Prolog
+
+## Introduction
+
+This library provides asyncronous semantics to Scryer Prolog users.
+
+## Limitations
+
+### Async, not Concurrent
+Note that Scryer Prolog is currently single-threaded, and so these tools provide asyncronicity but *not* concurrency. However, a form of concurrency can be achieved by combining this library with [`library(process)`](https://www.scryer.pl/process).
+
+### Limited options for non-blocking I/O
+Right now, there are few options available for non-blocking I/O. Some gymnastics with `library(process)` are required. This will change in the near future.
+
+## Use Cases
+
+There are certain concepts which map well to asyncronous semantics, such as managing multiple I/O connections to external programs or anything that may need to operate in a "persistent loop" while the rest of the program continues running.
+
+For instance, we could imagine:
+
+```prolog
+
+server :-
+    server_options(connection_timeout_ms(Timeout)),
+    wait_for_connection(Con, Timeout),
+    handle_connection(Con),
+    server.
+
+```
+
+Here, the best implementation of `handle_connection/1` is non-obvious. A signicant amount of machinery may be required, possibly involving undesirable mutational operations, in order to make the predicate non-blocking. Otherwise, `handle_connection/1` will block until the connection terminates, preventing any other connections from accessing the system.
+
+However, with a simple adjustment:
+
+```prolog
+server_async :-
+    server_options(connection_timeout_ms(Timeout)),
+    wait_for_connection(Con, Timeout),
+    await(handle_connection_async(Con)),
+    await(server_async).
+
+
+main :-
+    async_event_loop([server]).
```

Comment:
Should this be the following?

```suggestion
    async_event_loop([server_async]).
```